### ISSUE TITLE
Question on validating table data in winforms

### ISSUE THREAD

**Issue by muruga85** [2gis/Winium.Desktop#134]
Hi @skyline-gleb  and @NickAb ... Can you please let me know if there is a method in Winium to validate the table rows? I have an object of ControlType 'table' with 20 rows and I need 2nd column text of all 20 rows. Can this be accomplished?

**Comment by skyline-gleb** [comment]
Hi! For your case needed https://github.com/2gis/Winium.Elements (dotnet and java only)

**Comment by muruga85** [comment]
Hi @skyline-gleb .. I've already tried using Winium.Elements.Desktop namespace but that didnt work out. This control type is not being taken as Data grid as well. When I tried using DataGrid, it showed 'NO GET TEXT' error message. Please advise

**Comment by skyline-gleb** [comment]
More info please. For example, screenshot your table from UISpy

**Comment by muruga85** [comment]
Please see below:
ControlType: "ControlType.Table
AutomationID: AA
LocalizedControlType: "Table"

If we expand this, it will display all the 20 rows it with ControlType as "Custom". Please advise

**Comment by muruga85** [comment]
@skyline-gleb .. Can you please respond to this query?

**Comment by muruga85** [comment]
@skyline-gleb  and @NickAb - Can you please respond to this query at the earliest?

**Comment by muruga85** [comment]
@NickAb  - Eagerly waiting for the response.

**Comment by NickAb** [comment]
Could you please provide screenshot from UISpy that shows table expanded, so we can understand structure of table elements?

**Comment by muruga85** [comment]
ControlType: "ControlType.Table
AutomationID: AA
LocalizedControlType: "Table"

If we expand this, it will display all the 20 rows it with ControlType as "Custom". Please advise

**Comment by muruga85** [comment]
If we expand this, it will display all the 20 rows it with ControlType as "Custom". Please advise

**Comment by NickAb** [comment]
In you first comment you mentioned second column, but you don't say how they are nested in this `ControlType` `Custom` rows.

**Comment by muruga85** [comment]
@NickAb  - When you expand the Rows, it will have all the columns listed below that. Even this has the control type as 'Custom'. All I want is, just the Name attribute of this column. Please advise

**Comment by muruga85** [comment]
@NickAb - Do you require any additional information?

**Comment by NickAb** [comment]
You can find table by AutomationId and then search starting from the table using XPath.
```py
table.find_elements_by_xpath("/*[@ControlType='Custom']/*[@ControlType='Custom'][2]")
```
You might need to tune this expression.

Other option would be to first find all rows simple by searching by `ControlType='Custom'` from table.
Then iterate search results (rows) and for each search again by `ControlType='Custom'` and take the second result on each iteration.

**Comment by muruga85** [comment]
No @skyline-gleb .. I tried working out the simplest solution provided by you (Option 2)  but didnt work out. All of them returned no rows. Please advise.

```cs
Table.FindElements(By.XPath("/*[contains(@ControlType, 'ControlType.Custom')]"))
Table.FindElements(By.XPath("/*[contains(@ControlType, 'Custom') and contains(@Name, 'Row')]"))
```

**Comment by NickAb** [comment]
Does `ControlType` equals to `Custom` or `ControlType.Custom`?

Have you tried simple iterating over rows? Assuming `Table` is a table with *AutomationID: AA* as stated above by you. Iteration would look something like that:

```cs
var rows = Table.FindElements(By.Xpath("/*"));

foreach(var row in rows)
{
    var secondCell = row.FindElements(By.XPath("/*")[1]);
   // ...
}

```

**Comment by muruga85** [comment]
ControlType equals ControlType.Custom in UISPY. And moreover, I'm finding the table using XPath : /*[contains(@ControlType, 'ControlType.Table')] or with Automation ID AA.

**Comment by NickAb** [comment]
Try `/*` from `Table` and see if it will find any elements, maybe there some problem with locating any elements in the table. What control do use for table? Is it standard WinForms element or some custom element?

**Comment by muruga85** [comment]
I tried iterating as mentioned by you simply with /* xpath, it gives 11 rows but the table actually has more than 11 rows. When I try to print row.text with in foreach loop its gives 'NO GET TEXT'.

Table is standard winforms element with ControlType as 'ControlType.Table'

**Comment by NickAb** [comment]
- How many rows you can see on screen at any given moment? Is it 11 or less? This might be caused by virtual mode and cell reuse. Basically table has only small subset of rows needed to fill visible areas, all other rows are virtual and do not render until you scroll your table to make these rows visible. You can use two identical sequential calls to `Table.FindElements` to check this. Put breakpoint in your tests after the first call, then scroll the table manually and step to the next call, it should return similar number of rows as the first call, but some rows will be different, depending on how much you scrolled. This is only a guess, so please check this out.

- `row` might not have text pattern implemented, you will need to call `.Text` on some child element in a row, that child element should be an element that does have text itself.

**Comment by muruga85** [comment]
I can see 12 rows on the screen but on the first call I did only 5 rows returned. I did manually scroll the table then and made the second call and found only 5 rows returned but actually there are more rows on the screen.

Even .Text on some child element throws an error 'NO GET TEXT' 

Totally struck up with this table validation.

**Comment by muruga85** [comment]
Hello @NickAb  - Finally I was able to resolve the issue and successfully validated the row values. Below is the xpath used to identify rows under the table:

("./*[contains(@ControlType, 'ControlType.Custom')]")

Only thing which we were missing is just the dot before forward slash. Then I went ahead and iterated through each rows to find the status values.

I know this was was taking quite a long time but finally was able to resolve the issue successfully. Thanks for your guidance
